packages/ra-core/src/form/useResetSubmitErrors.ts

```diff
@@ -0,0 +1,25 @@
+import { useRef } from 'react';
+import { useForm, useFormState } from 'react-final-form';
+
+/**
+ * Reset the submission error when the corresponding field changes.
+ * final-form doesn't do this by default.
+ */
+const useResetSubmitErrors = () => {
+    const { mutators } = useForm();
+    const { values } = useFormState({ subscription: { values: true } });
```

Comment:
I find it weird that you need to call `useFormState` twice, especially since you've already called `useForm`. Maybe that's why you had to update the `FormWithRedirect` unit test.

I assume you could do the same with  `form.subscribe()`, without even calling `useFormState` once. 
Could you try that and see if it removes the extra render?